File: backend/routes/competencies.ts
import express from 'express';
import { competencyService } from '../services/competency-service.ts';
import { validateUUID } from '../utils/validation.ts';

const router = express.Router();

// Middleware to check authentication from Bearer token
async function requireAuth(req: express.Request, res: express.Response, next: express.NextFunction) {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '');
    if (!token) {
      return res.status(401).json({ success: false, error: 'Authentication required' });
    }

    const { sessionRepository } = await import('../db/repositories/session-repository.ts');
    const session = await sessionRepository.findByToken(token);

    if (!session) {
      return res.status(401).json({ success: false, error: 'Invalid or expired session' });
    }

    const { userService } = await import('../services/user-service.ts');
    const user = await userService.getUserById(session.user_id);
    if (!user) {
      return res.status(401).json({ success: false, error: 'User not found' });
    }

    (req as any).user = user;
    next();
  } catch (error) {
    console.error('Auth error:', error);
    return res.status(401).json({ success: false, error: 'Authentication failed' });
  }
}

// Middleware to check admin role
function requireAdmin(req: express.Request, res: express.Response, next: express.NextFunction) {
  const user = (req as any).user;
  if (!user || (user.role !== 'super_admin' && user.role !== 'company_admin')) {
    return res.status(403).json({ success: false, error: 'Admin access required' });
  }
  next();
}

// GET /api/competencies/industry - Get industry metadata and competencies (no auth required)
router.get('/industry', async (req: express.Request, res: express.Response) => {
  try {
    const { readJSONFile } = await import('../utils/file-storage.ts');

    // Load industry competencies and metadata from shared/data
    const industryCompetencies = readJSONFile('industry-competencies.json');
    const industryMetadata = readJSONFile('industry-metadata.json');

    return res.json({ success: true, data: { industryCompetencies, industryMetadata } });
  } catch (error) {
    console.error('Get industry metadata error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get industry metadata'
    });
  }
});

// GET /api/competencies - Get all competencies or filtered (no auth required for basic access)
router.get('/', async (req: express.Request, res: express.Response) => {
  try {
    const industry = req.query.industry as string;
    const category = req.query.category as string;

    let competencies;

    if (industry) {
      competencies = await competencyService.getCompetenciesByIndustry(industry);
    } else if (category) {
      competencies = await competencyService.getCompetenciesByCategory(category);
    } else {
      competencies = await competencyService.getAllCompetencies();
    }

    return res.json({ success: true, competencies });
  } catch (error) {
    console.error('Get competencies error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get competencies'
    });
  }
});

// POST /api/competencies - Create new competency (admin only)
router.post('/', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const { name, description, category, weight, industry } = req.body;

    // Validate required fields
    if (!name || !category) {
      return res.status(400).json({
        success: false,
        error: 'Missing required fields: name, category'
      });
    }

    // Validate weight if provided
    if (weight !== undefined && (weight < 0 || weight > 100)) {
      return res.status(400).json({
        success: false,
        error: 'Weight must be between 0 and 100'
      });
    }

    // Create competency
    const competency = await competencyService.createCompetency({
      name,
      description: description || null,
      category,
      weight: weight || 10,
      industry: industry || null,
    });

    if (!competency) {
      return res.status(400).json({ success: false, error: 'Failed to create competency' });
    }

    return res.status(201).json({ success: true, competency });
  } catch (error) {
    console.error('Create competency error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to create competency'
    });
  }
});

// GET /api/competencies/:id - Get competency by ID
router.get('/:id', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const competencyId = req.params.id;

    // Validate UUID
    if (!validateUUID(competencyId)) {
      return res.status(400).json({ success: false, error: 'Invalid competency ID format' });
    }

    // Get competency
    const competency = await competencyService.getCompetencyById(competencyId);

    if (!competency) {
      return res.status(404).json({ success: false, error: 'Competency not found' });
    }

    return res.json({ success: true, competency });
  } catch (error) {
    console.error('Get competency error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get competency'
    });
  }
});

// PUT /api/competencies/:id - Update competency (admin only)
router.put('/:id', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const competencyId = req.params.id;

    // Validate UUID
    if (!validateUUID(competencyId)) {
      return res.status(400).json({ success: false, error: 'Invalid competency ID format' });
    }

    const { name, description, category, weight, industry } = req.body;

    // Validate weight if provided
    if (weight !== undefined && (weight < 0 || weight > 100)) {
      return res.status(400).json({
        success: false,
        error: 'Weight must be between 0 and 100'
      });
    }

    // Update competency
    const updatedCompetency = await competencyService.updateCompetency(competencyId, {
      name,
      description,
      category,
      weight,
      industry,
    });

    if (!updatedCompetency) {
      return res.status(404).json({ success: false, error: 'Competency not found' });
    }

    return res.json({ success: true, competency: updatedCompetency });
  } catch (error) {
    console.error('Update competency error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to update competency'
    });
  }
});

// DELETE /api/competencies/:id - Delete competency (admin only)
router.delete('/:id', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const competencyId = req.params.id;

    // Validate UUID
    if (!validateUUID(competencyId)) {
      return res.status(400).json({ success: false, error: 'Invalid competency ID format' });
    }

    // Delete competency
    const success = await competencyService.deleteCompetency(competencyId);

    if (!success) {
      return res.status(404).json({ success: false, error: 'Competency not found' });
    }

    return res.json({ success: true, message: 'Competency deleted successfully' });
  } catch (error) {
    console.error('Delete competency error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to delete competency'
    });
  }
});

export default router;

File: backend-python/app/routers/competencies.py
"""
Competencies Router
Competency management endpoints
"""

import logging
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, status, Depends, Body

from app.services.competency_service import competency_service
from app.services.industry_service import industry_service
from app.middleware.auth import get_current_user, require_admin
from app.models.user import UserData
from app.utils.validation import validate_uuid

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/competencies", tags=["Competencies"])


@router.get("/industry")
async def get_industry_metadata():
    """Get industry metadata and competencies (no auth required)."""
    try:
        industry_competencies = await industry_service.get_industry_competencies()
        industry_metadata = await industry_service.get_industry_metadata()

        return {
            "success": True,
            "data": {
                "industryCompetencies": industry_competencies,
                "industryMetadata": industry_metadata,
            },
        }
    except Exception as e:
        logger.error(f"Get industry metadata error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("")
async def get_competencies(
    industry: str | None = None,
    category: str | None = None,
):
    """Get all competencies or filtered (no auth required for basic access)."""
    try:
        if industry:
            competencies = await competency_service.get_competencies_by_industry(industry)
        elif category:
            competencies = await competency_service.get_competencies_by_category(category)
        else:
            competencies = await competency_service.get_all_competencies()

        return {"success": True, "competencies": [c.model_dump() for c in competencies]}
    except Exception as e:
        logger.error(f"Get competencies error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_competency(
    admin: Annotated[UserData, Depends(require_admin)],
    name: str = Body(...),
    category: str = Body(...),
    description: str | None = Body(None),
    weight: int | None = Body(None),
    industry: str | None = Body(None),
):
    """Create new competency (admin only)."""
    try:
        # Validate required fields
        if not name or not category:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Missing required fields: name, category",
            )

        # Validate weight if provided
        if weight is not None and (weight < 0 or weight > 100):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Weight must be between 0 and 100",
            )

        competency = await competency_service.create_competency({
            "name": name,
            "description": description,
            "category": category,
            "weight": weight or 10,
            "industry": industry,
        })

        if not competency:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Failed to create competency")

        return {"success": True, "competency": competency.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Create competency error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/{competency_id}")
async def get_competency(
    competency_id: str,
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get competency by ID."""
    try:
        if not validate_uuid(competency_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid competency ID format")

        competency = await competency_service.get_competency_by_id(competency_id)

        if not competency:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Competency not found")

        return {"success": True, "competency": competency.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get competency error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/{competency_id}")
async def update_competency(
    competency_id: str,
    admin: Annotated[UserData, Depends(require_admin)],
    name: str | None = Body(None),
    description: str | None = Body(None),
    category: str | None = Body(None),
    weight: int | None = Body(None),
    industry: str | None = Body(None),
):
    """Update competency (admin only)."""
    try:
        if not validate_uuid(competency_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid competency ID format")

        # Validate weight if provided
        if weight is not None and (weight < 0 or weight > 100):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Weight must be between 0 and 100",
            )

        update_data: dict[str, Any] = {}
        if name is not None:
            update_data["name"] = name
        if description is not None:
            update_data["description"] = description
        if category is not None:
            update_data["category"] = category
        if weight is not None:
            update_data["weight"] = weight
        if industry is not None:
            update_data["industry"] = industry

        updated_competency = await competency_service.update_competency(competency_id, update_data)

        if not updated_competency:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Competency not found")

        return {"success": True, "competency": updated_competency.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update competency error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.delete("/{competency_id}")
async def delete_competency(
    competency_id: str,
    admin: Annotated[UserData, Depends(require_admin)],
):
    """Delete competency (admin only)."""
    try:
        if not validate_uuid(competency_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid competency ID format")

        success = await competency_service.delete_competency(competency_id)

        if not success:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Competency not found")

        return {"success": True, "message": "Competency deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Delete competency error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
